File: app/api/routers/cameras.py
"""FastAPI router for camera management endpoints.

Exposes CRUD operations, connection testing, and on-demand snapshot
triggering for cameras.
"""

from fastapi import APIRouter, Depends, HTTPException
from loguru import logger

from app.api.deps import get_camera_service, get_snapshot_service
from app.application.services.camera_service import CameraService
from app.application.services.snapshot_service import SnapshotService
from app.domain.schemas import (
    CameraCreate,
    CameraRead,
    CameraUpdate,
    CameraTestResult,
    CameraWithLastSnapshot,
    SnapshotForceResult,
)
from app.scheduler import remove_camera_job, reschedule_camera, schedule_camera

router = APIRouter(prefix="/api/cameras", tags=["cameras"])


@router.get("", response_model=list[CameraWithLastSnapshot])
async def list_cameras(
    service: CameraService = Depends(get_camera_service),
    snap_service: SnapshotService = Depends(get_snapshot_service),
):
    """List all cameras with their latest snapshot.

    \f
    Returns dashboard data for every camera, including the most recent
    snapshot (if any) captured for each.
    """
    data = await snap_service.get_dashboard_data()
    logger.debug(f"List cameras returned {len(data)} entries")
    return data


@router.get("/{camera_id}", response_model=CameraRead)
async def get_camera(
    camera_id: int,
    service: CameraService = Depends(get_camera_service),
):
    """Retrieve a single camera by its identifier.

    \f
    Args:
        camera_id: The unique identifier of the camera.

    Returns:
        The serialized camera record.

    Raises:
        HTTPException: 404 if the camera does not exist.
    """
    camera = await service.get_camera(camera_id)
    if not camera:
        logger.warning(f"Camera {camera_id} not found (GET)")
        raise HTTPException(status_code=404, detail="Camera not found")
    return CameraRead.model_validate(camera)


@router.post("", response_model=CameraRead, status_code=201)
async def create_camera(
    data: CameraCreate,
    service: CameraService = Depends(get_camera_service),
):
    """Create a new camera and schedule it if enabled.

    \f
    Validates the input payload, persists a new camera record, and
    immediately schedules the capture job when the camera is enabled.

    Args:
        data: The camera creation payload.
    """
    camera = await service.create_camera(data)
    await service.commit()
    if camera.enabled:
        schedule_camera(camera.id, camera.interval_seconds)
    logger.info(f"Camera created via API: {camera.name} (id={camera.id})")
    return CameraRead.model_validate(camera)


@router.put("/{camera_id}", response_model=CameraRead)
async def update_camera(
    camera_id: int,
    data: CameraUpdate,
    service: CameraService = Depends(get_camera_service),
):
    """Update an existing camera and reschedule its capture job.

    \f
    Applies the supplied field updates to the camera, persists them, and
    immediately updates the scheduler: enabled cameras are rescheduled
    with their current interval, disabled cameras have their job removed.

    Args:
        camera_id: The unique identifier of the camera.
        data: The camera update payload (partial update).

    Returns:
        The serialized updated camera record.

    Raises:
        HTTPException: 404 if the camera does not exist.
    """
    camera = await service.update_camera(camera_id, data)
    if not camera:
        logger.warning(f"Camera {camera_id} not found (PUT)")
        raise HTTPException(status_code=404, detail="Camera not found")
    await service.commit()
    if camera.enabled:
        reschedule_camera(camera.id, camera.interval_seconds)
    else:
        remove_camera_job(camera.id)
    logger.info(f"Camera {camera_id} updated via API")
    return CameraRead.model_validate(camera)


@router.delete("/{camera_id}", status_code=204)
async def delete_camera(
    camera_id: int,
    service: CameraService = Depends(get_camera_service),
):
    """Delete a camera and remove its scheduled job.

    \f
    Removes the camera record permanently and stops any scheduled capture
    job for it.

    Args:
        camera_id: The unique identifier of the camera.

    Raises:
        HTTPException: 404 if the camera does not exist.
    """
    deleted = await service.delete_camera(camera_id)
    if not deleted:
        logger.warning(f"Camera {camera_id} not found (DELETE)")
        raise HTTPException(status_code=404, detail="Camera not found")
    await service.commit()
    remove_camera_job(camera_id)
    logger.info(f"Camera {camera_id} deleted via API")


@router.post("/test", response_model=CameraTestResult)
async def test_camera(data: CameraCreate):
    """Test an ONVIF camera connection without persisting it.

    \f
    Attempts to connect to the camera using the supplied credentials and
    reports reachability plus discovered media profiles.

    Args:
        data: The camera payload containing host, port, and credentials.

    Returns:
        A test result with reachability status, profile tokens, and any
        error message.
    """
    from app.infrastructure.onvif import ONVIFCameraClient
    onvif = ONVIFCameraClient()
    logger.info(f"Testing camera {data.host}:{data.port} via API")
    reachable, profiles, error = onvif.test_connection(
        data.host, data.port, data.username, data.password
    )
    return CameraTestResult(reachable=reachable, profiles=profiles, error=error)


@router.post("/{camera_id}/snapshot", response_model=SnapshotForceResult)
async def force_snapshot(
    camera_id: int,
    service: SnapshotService = Depends(get_snapshot_service),
    camera_service: CameraService = Depends(get_camera_service),
):
    """Force an immediate snapshot capture for a camera.

    \f
    Triggers an out-of-schedule capture and returns a result summary
    including the image path on success.

    Args:
        camera_id: The unique identifier of the camera.

    Returns:
        A force-snapshot result with success flag and metadata.

    Raises:
        HTTPException: 404 if the camera does not exist.
    """
    snapshot = await service.force_capture(camera_id)
    if not snapshot:
        logger.warning(f"Force snapshot: camera {camera_id} not found")
        raise HTTPException(status_code=404, detail="Camera not found")
    camera = await camera_service.get_camera(camera_id)
    logger.info(f"Force snapshot completed for camera {camera_id}: status={snapshot.status}")
    return SnapshotForceResult(
        camera_id=snapshot.camera_id,
        camera_name=camera.name if camera else "",
        success=snapshot.status == "success",
        image_path=snapshot.image_path if snapshot.status == "success" else None,
        error=snapshot.error_message,
    )

File: internal/api/cameras.go
package api

import (
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/rs/zerolog/log"

	"github.com/Lmex89/home-cameras/internal/domain"
)

// camerasRoutes registers the /api/cameras endpoints (parity with the
// legacy cameras.py router).
//
// Args:
//
//	r: The chi subrouter.
func (s *Server) camerasRoutes(r chi.Router) {
	r.Get("/", s.listCameras)
	r.Post("/", s.createCamera)
	r.Post("/test", s.testCamera)
	r.Route("/{cameraID}", func(r chi.Router) {
		r.Get("/", s.getCamera)
		r.Put("/", s.updateCamera)
		r.Delete("/", s.deleteCamera)
		r.Post("/snapshot", s.forceSnapshot)
		r.Get("/stream", s.streamCamera)
	})
}

// listCameras returns all cameras with dashboard data (latest snapshot
// and total count). GET /api/cameras
func (s *Server) listCameras(w http.ResponseWriter, r *http.Request) {
	data, err := s.deps.SnapshotSvc.GetDashboardData(r.Context())
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	writeJSON(w, http.StatusOK, data)
}

// getCamera returns a single camera. GET /api/cameras/{cameraID}
func (s *Server) getCamera(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "cameraID")
	if !ok {
		return
	}
	cam, err := s.deps.CameraSvc.Get(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "Camera not found")
		return
	}
	writeJSON(w, http.StatusOK, cam)
}

// createCamera validates and persists a new camera. POST /api/cameras
func (s *Server) createCamera(w http.ResponseWriter, r *http.Request) {
	var payload domain.CameraCreate
	if err := readJSON(r, &payload); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	// Input validation mirroring the Pydantic constraints.
	if trimSpaces(payload.Name) == "" {
		writeError(w, http.StatusBadRequest, "name is required")
		return
	}
	if payload.CameraType == "" {
		payload.CameraType = "ip"
	}
	if payload.CameraType != "ip" && payload.CameraType != "usb" {
		writeError(w, http.StatusBadRequest, "camera_type must be 'ip' or 'usb'")
		return
	}
	if payload.CameraType == "usb" {
		if payload.DevicePath == nil || trimSpaces(*payload.DevicePath) == "" {
			writeError(w, http.StatusBadRequest, "device_path is required for USB cameras")
			return
		}
	} else {
		if trimSpaces(payload.Host) == "" {
			writeError(w, http.StatusBadRequest, "host is required for IP cameras")
			return
		}
		if payload.Port < 1 || payload.Port > 65535 {
			writeError(w, http.StatusBadRequest, "port must be between 1 and 65535")
			return
		}
	}
	if payload.IntervalSeconds == 0 {
		payload.IntervalSeconds = s.deps.Cfg.DefaultIntervalSeconds
	}
	if payload.IntervalSeconds < 10 {
		writeError(w, http.StatusBadRequest, "interval_seconds must be >= 10")
		return
	}

	cam, err := s.deps.CameraSvc.Create(r.Context(), payload)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	if cam.Enabled {
		s.deps.Sched.ScheduleCamera(cam.ID, cam.IntervalSeconds, time.Now())
	}
	log.Info().Int64("id", cam.ID).Str("name", cam.Name).Msg("camera created via API")
	writeJSON(w, http.StatusCreated, cam)
}

// updateCamera applies a partial update and reschedules the capture
// job. PUT /api/cameras/{cameraID}
func (s *Server) updateCamera(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "cameraID")
	if !ok {
		return
	}
	var payload domain.CameraUpdate
	if err := readJSON(r, &payload); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	cam, err := s.deps.CameraSvc.Update(r.Context(), id, payload)
	if err != nil {
		mapRepoErr(w, err, "Camera not found")
		return
	}
	if cam.Enabled {
		s.deps.Sched.RescheduleCamera(cam.ID, cam.IntervalSeconds)
	} else {
		s.deps.Sched.RemoveCamera(cam.ID)
	}
	log.Info().Int64("id", id).Msg("camera updated via API")
	writeJSON(w, http.StatusOK, cam)
}

// deleteCamera removes a camera and its capture job. DELETE /api/cameras/{cameraID}
func (s *Server) deleteCamera(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "cameraID")
	if !ok {
		return
	}
	deleted, err := s.deps.CameraSvc.Delete(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	if !deleted {
		writeError(w, http.StatusNotFound, "Camera not found")
		return
	}
	s.deps.Sched.RemoveCamera(id)
	log.Info().Int64("id", id).Msg("camera deleted via API")
	w.WriteHeader(http.StatusNoContent)
}

// testCamera probes an ONVIF endpoint without persisting. POST /api/cameras/test
func (s *Server) testCamera(w http.ResponseWriter, r *http.Request) {
	var payload domain.CameraCreate
	if err := readJSON(r, &payload); err != nil {
		writeError(w, http.StatusBadRequest, err.Error())
		return
	}
	log.Info().Str("host", payload.Host).Int("port", payload.Port).Msg("testing camera via API")
	result, err := s.deps.CameraSvc.Test(r.Context(), payload.Host, payload.Port, payload.Username, payload.Password)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, result)
}

// forceSnapshot triggers an immediate out-of-schedule capture.
// POST /api/cameras/{cameraID}/snapshot
func (s *Server) forceSnapshot(w http.ResponseWriter, r *http.Request) {
	id, ok := pathInt64(w, r, "cameraID")
	if !ok {
		return
	}
	cam, err := s.deps.CameraSvc.Get(r.Context(), id)
	if err != nil {
		mapRepoErr(w, err, "Camera not found")
		return
	}
	snap, err := s.deps.SnapshotSvc.Capture(r.Context(), *cam)
	if err != nil {
		mapRepoErr(w, err, "")
		return
	}
	result := domain.SnapshotForceResult{
		CameraID:   snap.CameraID,
		CameraName: cam.Name,
		Success:    snap.Status == "success",
		Error:      snap.ErrorMessage,
	}
	if snap.Status == "success" {
		result.ImagePath = strPtr(snap.ImagePath)
	}
	log.Info().Int64("camera_id", id).Str("status", snap.Status).Msg("force snapshot completed")
	writeJSON(w, http.StatusOK, result)
}
